timemachine/fe/rbfe.py

```diff
@@ -0,0 +1,423 @@
+    all_U_fns = []
+    for bp in bps:
+
+        def batch_U_fn(xs, boxes, bp_impl):
+            Us = []
+            for x, box in zip(xs, boxes):
+                # tbd optimize to "selective" later
+                _, _, U = bp_impl.execute(x, box, lamb)
+                Us.append(U)
+            return np.array(Us)
+
+        # extra functools.partial is needed to deal with closure jank
+        all_U_fns.append(functools.partial(batch_U_fn, bp_impl=bp))
+
+    return all_U_fns
+
+
+class HostConfig:
+    def __init__(self, omm_system, conf, box):
+        self.omm_system = omm_system
+        self.conf = conf
+        self.box = box
+
+
+def generate_samples(st, host_config, temperature, lambda_schedule, n_frames, seed):
+    """
+    Generate samples given a lambda_schedule. Let L be the number
+    of lambda windows T be the number of frames.
+
+    Parameters
+    ----------
+    st: SingleTopology
+        Topology container for the alchemical ligand
+
+    host_config: HostConfig
+        Configuration of the host
+
+    temperature: float
+        Temperature of the system, in Kelvins
+
+    lambda_schedule: np.array(L,)
+        Sequence of lambda windows
+
+    n_frames: int
+        Number of samples to generate per window
+
+    Returns
+    -------
+    3-tuple of all_frames, all_boxes, all_U_fns
+        all_frames has shape (L,T,N,3), all_boxes has shape (L,T,3,3), all_U_fns has shape (L,)
+
+    """
+    host_bps, host_masses = openmm_deserializer.deserialize_system(host_config.omm_system, cutoff=1.2)
+    host_conf = minimizer.minimize_host_4d(
+        [st.mol_a, st.mol_b],
+        host_config.omm_system,
+        host_config.conf,
+        st.ff,
+        host_config.box,
+    )
+
+    all_U_fns = []
+    all_frames = []
+    all_boxes = []
+
+    for lamb_idx, lamb in enumerate(lambda_schedule):
+        hgs = st.combine_with_host(convert_bps_into_system(host_bps), lamb=lamb)
+        all_U_fns.append(hgs.get_U_fns())
+        bound_impls = [p.bound_impl(np.float32) for p in hgs.get_U_fns()]
+        # minimize water box around the ligand by 4D-decoupling
+        mol_a_conf = get_romol_conf(st.mol_a)
+        mol_b_conf = get_romol_conf(st.mol_b)
+        ligand_conf = st.combine_confs(mol_a_conf, mol_b_conf)
+        combined_conf = np.concatenate([host_conf, ligand_conf])
+        x0 = combined_conf
+        v0 = np.zeros_like(x0)
+        box0 = host_config.box
+        group_idxs = get_group_indices(get_bond_list(hgs.bond))
+        run_seed = seed + lamb_idx
+        combined_masses = np.concatenate([host_masses, st.combine_masses()])
+        frames, boxes = sample(
+            bound_impls, combined_masses, temperature, x0, v0, box0, lamb, group_idxs, run_seed, n_frames=n_frames
+        )
+        all_frames.append(frames)
+        all_boxes.append(boxes)
+
+    return all_frames, all_boxes, all_U_fns
+
+
+def plot_BAR(df, df_err, fwd_delta_u, rev_delta_u, title, plot_idx):
```

Comment:
Should this be moved into a plotting utilities module or similar?

(Implementation does not seem RBFE-specific -- also seems equally applicable to other cases where forward and reverse works can be computed.)

Reply:
this is specific to this script since it's calling subplots (as the outside code sets-up the figure sizes etc.)

Reply:
per @mcwitt comment on using axes

(maybe if this took in arbitrary `plot` object (could be ax, or plot), as opposed to `plot_idx` object, this can be refactored out)